types/lib/linear.go

```diff
@@ -0,0 +1,245 @@
+package lib
+
+import (
+	"fmt"
+	"strconv"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	wire "github.com/cosmos/cosmos-sdk/wire"
+)
+
+// Linear defines a primitive mapper type
+type Linear struct {
+	cdc   *wire.Codec
+	store sdk.KVStore
+}
+
+// List is a Linear interface that provides list-like functions
+// It panics when the element type cannot be (un/)marshalled by the codec
+type List interface {
+
+	// Len() returns the length of the list
+	// The length is only increased by Push() and not decreased
+	// List dosen't check if an index is in bounds
+	// The user should check Len() before doing any actions
+	Len() uint64
+
+	// Get() returns the element by its index
+	Get(uint64, interface{}) error
+
+	// Set() stores the element to the given position
+	// Setting element out of range will break length counting
+	// Use Push() instead of Set() to append a new element
+	Set(uint64, interface{})
+
+	// Delete() deletes the element in the given position
+	// Other elements' indices are preserved after deletion
+	// Panics when the index is out of range
+	Delete(uint64)
+
+	// Push() inserts the element to the end of the list
+	// It will increase the length when it is called
+	Push(interface{})
+
+	// Iterate*() is used to iterate over all existing elements in the list
+	// Return true in the continuation to break
+	// The second element of the continuation will indicate the position of the element
+	// Using it with Get() will return the same one with the provided element
+
+	// CONTRACT: No writes may happen within a domain while iterating over it.
+	Iterate(interface{}, func(uint64) bool)
+}
+
+// NewList constructs new List
+func NewList(cdc *wire.Codec, store sdk.KVStore) List {
+	return Linear{
+		cdc:   cdc,
+		store: store,
+	}
+}
+
+// Key for the length of the list
+func LengthKey() []byte {
```

Comment:
Why is this a function instead of a constant?

Reply:
`[]byte` cannot be declared as const, only as var, which can be modified from other malicious code.